src/gluonts/torch/util.py

```diff
@@ -0,0 +1,94 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+from typing import Optional, Type
+import inspect
+
+import torch
+
+
+def copy_parameters(
+    net_source: torch.nn.Module,
+    net_dest: torch.nn.Module,
+    strict: Optional[bool] = True,
+) -> None:
+    """
+    Copies parameters from one network to another.
+
+    Parameters
+    ----------
+    net_source
+        Input network.
+    net_dest
+        Output network.
+    strict:
+        whether to strictly enforce that the keys
+        in :attr:`state_dict` match the keys returned by this module's
+        :meth:`~torch.nn.Module.state_dict` function. Default: ``True``
+    """
+
+    net_dest.load_state_dict(net_source.state_dict(), strict=strict)
+
+
+def get_forward_input_names(module: Type[torch.nn.Module]):
+    params = inspect.signature(module.forward).parameters
+    param_names = [k for k, v in params.items() if not str(v).startswith("*")]
+    assert param_names[0] == "self", (
+        f"Expected first argument of forward to be `self`, "
+        f"but found `{param_names[0]}`"
+    )
+    return param_names[1:]  # skip: self
+
+
+def weighted_average(
+    x: torch.Tensor, weights: Optional[torch.Tensor] = None, dim=None
+) -> torch.Tensor:
+    """
+    Computes the weighted average of a given tensor across a given dim, masking
+    values associated with weight zero,
+    meaning instead of `nan * 0 = nan` you will get `0 * 0 = 0`.
+
+    Parameters
+    ----------
+    x
+        Input tensor, of which the average must be computed.
+    weights
+        Weights tensor, of the same shape as `x`.
+    dim
+        The dim along which to average `x`
+
+    Returns
+    -------
+    Tensor:
+        The tensor with values averaged along the specified `dim`.
+    """
+    if weights is not None:
+        weighted_tensor = torch.where(
+            weights != 0, x * weights, torch.zeros_like(x)
+        )
+        sum_weights = torch.clamp(
+            weights.sum(dim=dim) if dim else weights.sum(), min=1.0
+        )
+        return (
+            weighted_tensor.sum(dim=dim) if dim else weighted_tensor.sum()
+        ) / sum_weights
+    else:
+        return x.mean(dim=dim)
+
+
+class IterableDataset(torch.utils.data.IterableDataset):
```

Comment:
Is this really necessary? `torch`'s `IterableDataset` is really just `__iter__` + some `NotImplementedError`s raised when accessing `__len__` or `__getitem__`

Reply:
Exactly, and also `__iter__` is not defined there. Its docstring says:
```
All datasets that represent an iterable of data samples should subclass it.
```
so I did it 😄